lighthouse-core/report/html/renderer/element-screenshot-renderer.js

```diff
@@ -0,0 +1,259 @@
+ */
+'use strict';
+
+/* globals self */
+
+/** @typedef {import('./dom.js')} DOM */
+/** @typedef {LH.Artifacts.Rect} Rect */
+/** @typedef {{width: number, height: number}} Size */
+
+/**
+ * @param {number} value
+ * @param {number} min
+ * @param {number} max
+ */
+function clamp(value, min, max) {
+  if (value < min) return min;
+  if (value > max) return max;
+  return value;
+}
+
+/**
+ * @param {Rect} rect
+ */
+function getRectCenterPoint(rect) {
+  return {
+    x: rect.left + rect.width / 2,
+    y: rect.top + rect.height / 2,
+  };
+}
+
+class ElementScreenshotRenderer {
+  /**
+   * Given the location of an element and the sizes of the preview and screenshot,
+   * compute the absolute positions (in screenshot coordinate scale) of the screenshot content
+   * and the highlighted rect around the element.
+   * @param {Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   * @param {Size} screenshotSize
+   */
+  static getScreenshotPositions(
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords, screenshotSize) {
+    //
+    const elementRectCenter = getRectCenterPoint(elementRectInScreenshotCoords);
+
+    // Try to center clipped region.
+    const screenshotLeftVisibleEdge = clamp(
+      elementRectCenter.x - elementPreviewSizeInScreenshotCoords.width / 2,
+      0, screenshotSize.width - elementPreviewSizeInScreenshotCoords.width
+    );
+    const screenshotTopVisisbleEdge = clamp(
+      elementRectCenter.y - elementPreviewSizeInScreenshotCoords.height / 2,
+      0, screenshotSize.height - elementPreviewSizeInScreenshotCoords.height
+    );
+
+    return {
+      screenshot: {
+        left: screenshotLeftVisibleEdge,
+        top: screenshotTopVisisbleEdge,
+      },
+      clip: {
+        left: elementRectInScreenshotCoords.left - screenshotLeftVisibleEdge,
+        top: elementRectInScreenshotCoords.top - screenshotTopVisisbleEdge,
+      },
+    };
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {HTMLElement} mask
+   * @param {{left: number, top: number}} positionClip
+   * @param {LH.Artifacts.Rect} elementRectInScreenshotCoords
+   * @param {Size} elementPreviewSizeInScreenshotCoords
+   */
+  static renderClipPath(dom, mask, positionClip,
+      elementRectInScreenshotCoords, elementPreviewSizeInScreenshotCoords) {
+    // Normalize values between 0-1.
+    const top = positionClip.top / elementPreviewSizeInScreenshotCoords.height;
+    const bottom =
+      top + elementRectInScreenshotCoords.height / elementPreviewSizeInScreenshotCoords.height;
+    const left = positionClip.left / elementPreviewSizeInScreenshotCoords.width;
+    const right =
+      left + elementRectInScreenshotCoords.width / elementPreviewSizeInScreenshotCoords.width;
+
+    const clipId = `clip-${top}-${bottom}-${left}-${right}`;
```

Comment:
FWIW this id won't be unique, and I believe it'll always use the first found in the page, not the closest or whatever. It may not have any noticeable effect because they'll be identical clip paths, but some things like styling `display: none` on the first clip path may bleed through to all the other uses of it (that's what happened in #9139)

Reply:
what do:

1. just add a large random number to id
1. search for existing `clipId` in `dom` and increment a number until unique
1. search for existing `clipId` in `dom` and skip making `clipPathSvg` if one already exists

Reply:
may I introduce to you: `getUniqueSuffix2`

... lol there's issues with global scope. any ideas?

Reply:
> may I introduce to you: `getUniqueSuffix2`
> 
> ... lol there's issues with global scope. any ideas?

haha. I guess the modules are a lie. `Util.getUniqueSuffix` for both?